doc/modules/calibration.rst

```diff
@@ -74,10 +74,14 @@ by showing the number of samples in each predicted probability bin.
 
 .. currentmodule:: sklearn.linear_model
 
-:class:`LogisticRegression` returns well calibrated predictions by default as it has a
+:class:`LogisticRegression` is more likely to return well calibrated predictions by itself as it has a
 canonical link function for its loss, i.e. the logit-link for the :ref:`log_loss`.
-This leads to the so-called **balance property**, see [8]_ and
-:ref:`Logistic_regression`.
+In the unpenalized case, this leads to the so-called **balance property**, see [8]_ and :ref:`Logistic_regression`.
+In the plot above, data is generated according to a linear mechanism, which is
+consistent with the :class:`LogisticRegression` model (the model is 'well specified'),
+and the value of the regularization parameter `C` (chosen arbitrarily here) happens to be
+appropriate (neither too strong nor too low). As a consequence, this model returns
+accurate predictions from its `predict_proba` method.
```

Comment:
Actually this is not guaranteed:

- https://github.com/scikit-learn/scikit-learn/blob/2d3629ff5/sklearn/datasets/_samples_generator.py#L42

with the default value of `n_clusters_per_class=2`, the data is not even necessarily linearly separable.

I tried to edit `n_clusters_per_class=1` but that does not seem to improve the calibration of the LR model: since the training data is very small (intentionally), it can happen that for some dataset RNG seeds, all models are badly calibrated.

Furthermore, I find that the plot of the example does not really show a well calibrated LR model. It's quite over-confident to me (almost as off-diagonal as the RF model but in a mirror way).

In #28231 I show how to automatically tune C on the NLL via `LogisticRegressionCV` and it seems to improve calibration a bit but it's still not perfect. If we merge #28231, I suggest to update this paragraph as:

```suggestion
In the plot above, the :class:`LogisticRegressionCV` model shows the reliability
diagram the closest to the diagonal. When the model is well-specified, that is
if the data generative mechanism involves a linear function of the input features,
and when the dataset size is large enough, we can expect this model yield well
calibrated predictions.
```

Reply:
I am not even sure that "well-specified" LR model is required for asymptotic calibration in retrospect. The general for of the balance property discussed in https://github.com/scikit-learn/scikit-learn/pull/28171#discussion_r1463811368 should be valid even when the LR model is mispecified. What matters is that the train and test sets are large enough to be close to the asymptotic regime and that regularization strength has tuned accordingly (close enough to zero when approaching the asymptotic regime).

Reply:
One could debate the differences between necessary and sufficient conditions as well as what "well calibrated" means exactly. To keep it short and general enough, "sufficiently good specification" is required for "good calibration" with LR, as is large enough data to train on.

Reply:
I edited the suggestion above. Let me know what you think.